shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm

```diff
@@ -655,96 +668,227 @@ static bool ClipRRectContainsPlatformViewBoundingRect(const SkRRect& clip_rrect,
+                  ios_context,                                     //
+                  ((FlutterView*)flutter_view_.get()).pixelFormat  //
+      );
+    }
+    return;
+  }
+
+  auto latch = std::make_shared<fml::CountDownLatch>(1u);
+  platform_task_runner_->PostTask([&]() {
+    // Create Missing Layers
+    for (auto i = 0u; i < missing_layer_count; i++) {
+      CreateLayer(gr_context,                                      //
+                  ios_context,                                     //
+                  ((FlutterView*)flutter_view_.get()).pixelFormat  //
+      );
+    }
+    latch->CountDown();
+  });
+  latch->Wait();
+}
+
+/// Update the buffers and mutate the platform views in CATransaction on the platform thread.
+void FlutterPlatformViewsController::PerformSubmit(
+    const LayersMap& platform_view_layers,
+    const std::vector<SurfaceFrame::DeferredSubmit>& callbacks,
+    std::map<int64_t, EmbeddedViewParams>& current_composition_params,
+    const std::unordered_set<int64_t>& views_to_recomposite,
+    const std::vector<int64_t>& composition_order,
+    const std::vector<std::shared_ptr<FlutterPlatformViewLayer>>& unused_layers,
+    const std::vector<UIView*>& views_to_dispose) {
+  TRACE_EVENT0("flutter", "FlutterPlatformViewsController::SubmitFrame::CATransaction");
+
+  [CATransaction begin];
+
+  // Configure Flutter overlay views.
+  for (const auto& [key, layers] : platform_view_layers) {
+    for (const auto& layer_data : layers) {
+      layer_data.layer->UpdateViewState(flutter_view_,         //
+                                        layer_data.rect,       //
+                                        layer_data.view_id,    //
+                                        layer_data.overlay_id  //
+      );
+    }
+  }
+
+  // Dispose unused Flutter Views.
+  for (auto& view : views_to_dispose) {
+    [view removeFromSuperview];
+  }
+
+  // Composite Platform Views.
+  for (int64_t view_id : views_to_recomposite) {
+    CompositeWithParams(view_id, current_composition_params[view_id]);
+  }
+
+  // Present callbacks.
+  for (const auto& cb : callbacks) {
+    cb();
+  }
+
+  // Organize the layers by their z indexes.
+  auto active_composition_order = BringLayersIntoView(platform_view_layers, composition_order);
+
+  // If a layer was allocated in the previous frame, but it's not used in the current frame,
+  // then it can be removed from the scene.
+  RemoveUnusedLayers(unused_layers, composition_order, active_composition_order);
 
   // If the frame is submitted with embedded platform views,
   // there should be a |[CATransaction begin]| call in this frame prior to all the drawing.
   // If that case, we need to commit the transaction.
-  CommitCATransactionIfNeeded();
-  return did_submit;
+  [CATransaction commit];
 }
 
-void FlutterPlatformViewsController::BringLayersIntoView(LayersMap layer_map) {
+std::vector<int64_t> FlutterPlatformViewsController::BringLayersIntoView(
```

Comment:
Oof, this was tough to read and I wonder if this couldn't be made a whole lot simpler and faster.

For one, mutating the view hierarchy can be pretty slow because of layout updated triggered potentially. So you want to minimize them.

And, in cases where the hierarchy changes, you are making addSubview perform a reorder. No matter if there is a reorder or not. And it seems to be response to a performance issue discovered later.

I am assuming that the view hierarchy is the same but you just need to reorder. In that case, you just need two arrays of integers (of z order). You already have one in the `composition_order`. Build the second one using [superview.subviews indexOfObject:sub]`. Now cycle over the two arrays (asserting beforehand that they are the same size) and call `-[UIView exchangeSubviewAtIndex:withSubviewAtIndex:]` only of the order at the specific index is different. If the order is the same, nothing changes. And you minimize reorders.


```c++
for (size_t i = 0; i < composition_order.size(); i++) {
  if (desired_order[i] != composition_order[i]) {
    exchange...
  }
}

```

That should be it.

Reply:
One complication is that I believe this is also adding the new flutter views, but we can account for that by adding them onto the end first and then moving them maybe.

Reply:
I tried a few things here, the issue is that the overlay views (not the platform views) need to be sequenced after the flutter views and don't have an ID. Also since the view hierarchy isn't the same I'd need a more complicated diff.

I think we can fix this, but since I'm not really changing this code i'd rather leave it as is.